### ISSUE TITLE
painkiller crashes at startup

### ISSUE THREAD

**Issue by iiv3** [iXit/Mesa-3D#115]
The crash happens only when using Nine with iXit Mesa3D (git-b41d72c). If I use Nine with vanilla Mesa3D 10.6.2 release, the game loads fine. (debug build fails with assert, caused by missing implementation of "TEXBEM" opcode. That opcode is implemented in iXit.) 

The game is old enough to have a demo, but the one I found was using directx8. It's possible that the demo provided by steam is the same.

[This log](https://gist.github.com/iiv3/3683cae7dd47c880fe03) is created by the command:
`WINEDEBUG=+d3dadapter9,+d3d9 NINE_DEBUG=device wine steam.exe steam://rungameid/39530`

I have created a log with `NINE_DEBUG=all` but it is about 1,5MB raw. I will provide it, if you think it is needed.

**Comment by axeldavy** [comment]
Seems like there is some issue with device Reset after using Stateblocks

**Comment by axeldavy** [comment]
It would be useful to know the precise line.

Perhaps add more DEBUG lines in nine_state_reset

**Comment by iiv3** [comment]
I have to apologize. I had `CFLAGS` with `-O2` and configure kept it. If I set it to `-O0` some functions are not inlined and I get the following backtrace
```
Unhandled exception: page fault on read access to 0x3dccccd5 in 32-bit code (0x7d5a49e4).
Register dump:
 CS:0023 SS:002b DS:002b ES:002b FS:0063 GS:006b
 EIP:7d5a49e4 ESP:0033f474 EBP:0033f48c EFLAGS:00210206(  R- --  I   - -P- )
 EAX:3dcccccd EBX:7d73338c ECX:f754e700 EDX:00000400
 ESI:06c536e8 EDI:06c53ccc
Stack dump:
0x0033f474:  000002d0 06c536e8 0033f4bc 00000000
0x0033f484:  7c902b18 06c536e8 0033f4ac 7d5a4a80
0x0033f494:  3dcccccd 00000000 7c90489c 7d5f229f
0x0033f4a4:  7c902b18 00000000 0033f4ec 7d5a7c6d
0x0033f4b4:  7c900838 00000000 0033f4ec 7d594348
0x0033f4c4:  7c902b18 00000000 00000010 00000001
Backtrace:
=>0 0x7d5a49e4 NineUnknown_Unbind+0xa() in d3dadapter9.so.1 (0x0033f48c)
  1 0x7d5a4a80 _nine_bind+0x26() in d3dadapter9.so.1 (0x0033f4ac)
  2 0x7d5a7c6d nine_state_clear+0xc2() in d3dadapter9.so.1 (0x0033f4ec)
  3 0x7d575986 NineDevice9_Reset+0x9d() in d3dadapter9.so.1 (0x0033f51c)
0x7d5a49e4 NineUnknown_Unbind+0xa in d3dadapter9.so.1: movl     0x8(%eax),%eax
```
I tried to add an `assert(This);` and `DBG("something")` in the *_Unbind function, but I don't get assertion nor any new text output in the log (with NINE_DEBUG=all). The faulty instruction moves few bytes, so the changes are present.

I think it would be easier if I try to do a bissect.

**Comment by okias** [comment]
about not printed DBG, are u sure u processed according this wiki > https://wiki.ixit.cz/d3d9_debugging ?

**Comment by iiv3** [comment]
@okias, I think my initial post uses the env.vars proposed in that page.
On a second thought, the issue might be much more trivial, like forgotten "\n". Will test later.

Anyway,
The result of the bisect is:
```
81bca22e7d6bed032c4a28702f9cf4f3460e7351 is the first bad commit
commit 81bca22e7d6bed032c4a28702f9cf4f3460e7351
Author: Axel Davy <axel.davy@ens.fr>
Date:   Fri May 8 16:54:53 2015 +0200

    st/nine: Implement EvictManagedResources
    
    Signed-off-by: Axel Davy <axel.davy@ens.fr>

:040000 040000 2a4559adf20ef99e5f46652866d705d329d0563a 91c811c1e127505e5cda35a191bbf3c993b49351 M      src
```

Here is the output of `git bisect log`
``` 
git bisect start 
# good: [e31bce4041122cd00712b60b4dc1eae6486f6579] svga: silence warnings about unexpected shader type
git bisect good e31bce4041122cd00712b60b4dc1eae6486f6579
# bad: [b41d72c8fb07f7a28ede6d7259dace7250ec77f4] target/d3dadapter9: Return Windows like card names Add support for multiple cards and fill in Windows like card name, driver name and version info. Use fallback for unknown vendors and unknown card names.
git bisect bad b41d72c8fb07f7a28ede6d7259dace7250ec77f4
# bad: [c733c68cdb433a394759a429376cd35606442f28] st/nine: Rework rasterizer states
git bisect bad c733c68cdb433a394759a429376cd35606442f28
# good: [d4d06948cf98137a34946cdfaae3eab7dd6d963c] st/nine: Only update dirty rect for UpdateTexture
git bisect good d4d06948cf98137a34946cdfaae3eab7dd6d963c
# bad: [869df635ec04dffbc35946ae4f168390dd9e9726] st/nine: Fix GenerateMipSubLevels potential crash
git bisect bad 869df635ec04dffbc35946ae4f168390dd9e9726
# bad: [325a5ffc449526b56cfc2de0655c8326bfb2052e] st/nine: Implement TEXBEM,TEXBEML and BEM
git bisect bad 325a5ffc449526b56cfc2de0655c8326bfb2052e
# bad: [81bca22e7d6bed032c4a28702f9cf4f3460e7351] st/nine: Implement EvictManagedResources
git bisect bad 81bca22e7d6bed032c4a28702f9cf4f3460e7351
# good: [420a4e19436c79c9fc441d5be5cb24b5e87a45ca] st/nine: Track managed textures
git bisect good 420a4e19436c79c9fc441d5be5cb24b5e87a45ca
# first bad commit: [81bca22e7d6bed032c4a28702f9cf4f3460e7351] st/nine: Implement EvictManagedResources
```

**Comment by iiv3** [comment]
OK, I've located the bug.

My initial bisect was wrong. Revering the suspected commit did not fix the issue. I redid the bisect and this time I ended with iXit/Mesa-3D@325a5ffc449526b56cfc2de0655c8326bfb2052e :
```
commit 325a5ffc449526b56cfc2de0655c8326bfb2052e
Author: Tiziano Bacocco <tizbac2@gmail.com>
Date:   Fri Jan 30 20:10:38 2015 +0100
st/nine: Implement TEXBEM,TEXBEML and BEM
```

I couldn't revert that commit automatically, so I went to disable it manually. To my surprise gutting out `DECL_SPECIAL(TEXBEM)` did not solve the issue.

It was however solved by disabling only the new block in `src/gallium/state_trackers/nine/device9.c
` - the one with the big `switch` . Here are some important snippets.
```
2631:     DBG("Stage=%u Type=%u Value=%08x\n", Stage, Type, Value);
2632:     nine_dump_D3DTSS_value(DBG_FF, Type, Value);
2633: 
2634:     user_assert(Stage < Elements(state->ff.tex_stage), D3DERR_INVALIDCALL);
[...]
2654:+    case D3DTSS_BUMPENVLOFFSET:
2655:+        bumpmap_index = 4 * 8 + 2 * Stage + 1;
[...]
2662:+        state->bumpmap_vars[bumpmap_index] = Value;
```

In `src/gallium/state_trackers/nine/nine_state.h` you can find the following lines:
```
104: #define NINE_MAX_SAMPLERS 21 /* PS + DMAP + VS */
[...]
153:+    uint32_t bumpmap_vars[48];`
[...]
189: uint32_t tex_stage[NINE_MAX_SAMPLERS][(NINED3DTSS_COUNT + 31) / 32];
```

In short,  `Stage < 21`, but the new code assumes it can't go over 8. (`4*8+2*8=48`).
To confirm it, I took a look of the big full log and found:
```
nine:device9:SetTextureStageState: Stage=15 Type=1 Value=00000001
nine:nine_dump_D3DTSS_value: D3DTSS_COLOROP = DISABLE
nine:device9:SetTextureStageState: Stage=15 Type=2 Value=00000002
nine:nine_dump_D3DTSS_value: D3DTSS_COLORARG1 = TEXTURE
nine:device9:SetTextureStageState: Stage=15 Type=3 Value=00000001
nine:nine_dump_D3DTSS_value: D3DTSS_COLORARG2 = CURRENT
nine:device9:SetTextureStageState: Stage=15 Type=4 Value=00000001
nine:nine_dump_D3DTSS_value: D3DTSS_ALPHAOP = DISABLE
nine:device9:SetTextureStageState: Stage=15 Type=5 Value=00000002
nine:nine_dump_D3DTSS_value: D3DTSS_ALPHAARG1 = TEXTURE
nine:device9:SetTextureStageState: Stage=15 Type=6 Value=00000001
nine:nine_dump_D3DTSS_value: D3DTSS_ALPHAARG2 = CURRENT
nine:device9:SetTextureStageState: Stage=15 Type=7 Value=3dcccccd
nine:nine_dump_D3DTSS_value: D3DTSS_BUMPENVMAT00 = 0.100000
nine:device9:SetTextureStageState: Stage=15 Type=8 Value=00000000
nine:nine_dump_D3DTSS_value: D3DTSS_BUMPENVMAT01 = 0.000000
nine:device9:SetTextureStageState: Stage=15 Type=9 Value=00000000
nine:nine_dump_D3DTSS_value: D3DTSS_BUMPENVMAT10 = 0.000000
nine:device9:SetTextureStageState: Stage=15 Type=10 Value=3dcccccd
nine:nine_dump_D3DTSS_value: D3DTSS_BUMPENVMAT11 = 0.100000
nine:device9:SetTextureStageState: Stage=15 Type=11 Value=0000000f
nine:nine_dump_D3DTSS_value: D3DTSS_TEXCOORDINDEX = PASSTHRU 15
nine:device9:SetTextureStageState: Stage=15 Type=24 Value=00000000
```

Well, increasing the size of `bumpmap_vars[]` solved the crash.

I don't know what the code does, but I suspect that it would be better to use `NINE_MAX_SAMPLERS_PS` a lot more in the code. To check `Stage`, to calculate the `bumpmap_index` and of course to set the size of `bumpmap_vars[]`.

**Comment by okias** [comment]
should be fixed by Axel commit 19dad7c9b9ac23c61160340c4584abb278b0e044 , please retest.

Thank you for report and good detective work :+1:

**Comment by iiv3** [comment]
I expected that he'd mention the commit and close this issue.

I've tested the patch before it was pushed, so it's confirmed that it works and doesn't cause any regressions.
